### ISSUE TITLE
Minimum web api requirements (SlotContentMixin not populated in Safari 12 mobile)

### ISSUE THREAD

**Issue by Pickachu** [elix/elix#107]
hello! I'm thinking into using the lib to build a big universal production-ready web application.

Looking at the code and at the demos i was not able to figure out the minimum browser apis needed for the components to run.

I was able to run the sliding pages demo (https://component.kitchen/demos/slidingPages.html) on desktop evergreen browser fine (chrome, safari and firefox), but when trying to run into chrome (78.0) mobile and safari (12.2) mobile _swiping the screen has no effect. A suspect that it is because of the initial state of the components being not populated. (edited)_

Upon inspecting the elements on Safari inspector, apparently the [internal.state] looks like some values did not update:

Safari mobile
![image](https://user-images.githubusercontent.com/286164/68423801-801cc700-0181-11ea-8108-78c8a024386e.png)

Google chrome desktop
![image](https://user-images.githubusercontent.com/286164/68423994-d984f600-0181-11ea-8fc9-4f81191bd783.png)

So, i have some doubts:

- Which are the minimum Web Api requirements for the lib support?
- Is there any limitation with the library in regards to using mobile on devices besides availability of native API's?
- Are there community plans to maintain mobile support?

btw, awesome architectural job on the lib selecting classes, building the interoperability and abstracting responsibilities!

**Comment by JanMiksovsky** [comment]
Thanks for taking a look at Elix for your project.

We test and support Elix components on both Chrome for Android and Mobile Safari. Both of those mobile browsers have supported all the basic web component APIs for several years. There is some new work coming in Elix to support a new CSS mechanism for styling components that will require browser updates, but even once that work lands, the basic components should still work as expected on production browsers. Of course, we do occasionally see mobile bugs; if you see things that look off, by all means file additional bugs.

As I read your issue report above, this is expected behavior. As described in https://component.kitchen/blog/posts/hiding-internal-framework-methods-and-properties-from-web-component-apis, the `internal.state` symbol is generally only reachable via an import reference. That's by design, to avoid complications that arise when developers using a component come to depend on internals the component author would like the freedom to change.

The post describes a debug flag, `?elixdebug=true`, which you can add to a page URL when debugging to expose a console-accessible `state` property on all Elix components.

It looks like some part of your bug description may have been cut off after, "but when trying to run into chrome (78.0) mobile and safari (12.2) mobile." If you're seeing a user-visible issue on that SlidingPages demo page, please let me know. As far as I can tell, that demo works as expected for me in both Chrome for Android and Mobile Safari.

**Comment by Pickachu** [comment]
Ops, my bad! Probably my internet failed to edit the main part of issue, i've already updated there. 

The main problem is that swiping has no effect (perhaps something to do with touch events), all swipeable elements (carousel, sliding pages, etc) are not working for me.

I've created a small video here to demonstrate: https://youtu.be/p80Kb_f_BS4

thanks for the fast reply!

**Comment by JanMiksovsky** [comment]
Hmm, that's really weird. (Thanks for making the video, BTW.)

I don't know if I've got a device still running Mobile Safari 12.x; it looks like I've got 13.x everywhere. We did make some changes to our touch/trackpad gesture support this summer, and it's theoretically possible those changes are broken in 12.x. I'll see if I can find an older device; if I can't find a device with Safari 12.x, it's going to be really hard to debug this.

And you see the same problem on Chrome for Android 78? I do have a phone that's still running Chrome 77, and the various swiping component demos work for me on that device.

In the meantime, can you try https://component.kitchen/demos/swipeDemo.html in Mobile Safari? That's the simplest demo of touch support we have. Since you're familiar with opening pages in the Safari inspector, it'd be helpful to know if you see any errors there.

It'd also be helpful if you could try using the inspector to set a breakpoint in TouchSwipeMixin.js, e.g., line 38 (the beginning of the `touchstart` handler), to see if the touch events are being received at all.

**Comment by Pickachu** [comment]
Hello again Jan!

The swipeDemo.html is working correctly, after testing the touch events in the TouchSwipeMixin as you suggested seems that all events are firing normally. After following the call stack i found that the `items` property on State were null and that prevented the goLeft() call initiated from the SwipeRight gesture symbol.

After 15 min of debugging I think i found the issue. The breakpoint on line 43 does not pause:

![image](https://user-images.githubusercontent.com/286164/68439947-3bf0ed00-01a8-11ea-8af3-4b892d964e37.png)

The event listener for slotchange at the SlotContentMixin is not being called on my safari mobile (but is called on chrome desktop):

```javascript
slot.addEventListener('slotchange', async () => {
    ... code never executed ...
});

```

That might be preventing SlotContentMixin from populating the items property for the SingleSelectionMixin, and then preventing swipes.

- Perhaps is this some weird implementation of the slot web api? [Safari 12 seems to support it](https://caniuse.com/#search=slotchange)
- Or perhaps the e event listener is being bound to late in the render pipeline?

Also contacted my friends and the demos are working fine on mobile Safari 13, and latest chrome for Android.

Thanks again for the help! If we get out of ideas I think it's not a big deal to just ignore the issue, we do not have many users on iOS 12 now.

**Comment by JanMiksovsky** [comment]
Ah, good work, I think you've found the issue!

Webkit in Safari 12 and earlier had a [bug](https://bugs.webkit.org/show_bug.cgi?id=188752) that we found and filed in which the `slotchange` event did not fire for initial content. We used to have a workaround for that bug in SlotContentMixin. When Apple shipped their fix in Safari 13, we dropped our workaround so that we could treat `slotchange` in all browsers the same way. In retrospect, that may have been premature; it might have been better to wait longer for users to migrate to iOS 13.

We could theoretically restore the workaround in order to support Safari 12, although that comes at a cost of reintroducing complexity into our codebase. If a potential Elix customer had a hard requirement to support Safari 12, we'd consider that. But if you think the issue only affects a small number of your users, my preference would be to leave this as is.
